lib/fika/type_checker.ex

```diff
@@ -367,17 +359,21 @@ defmodule Fika.TypeChecker do
   end
 
   defp do_infer_args_without_name(env, args) do
-    Enum.reduce_while(args, {:ok, [], env}, fn arg, {:ok, type_acc, env} ->
+    Enum.reduce_while(args, {:ok, %T.ArgList{}, env}, fn arg, {:ok, acc, env} ->
       case infer_exp(env, arg) do
         {:ok, type, env} ->
           Logger.debug("Argument is type: #{type}")
-          {:cont, {:ok, type_acc ++ [type], env}}
+          {:cont, {:ok, %{acc | value: acc.value ++ [type]}, env}}
 
         error ->
           Logger.debug("Argument cannot be inferred")
           {:halt, error}
       end
     end)
+    |> case do
+      {:ok, inferred, env} -> {:ok, to_string(inferred), env}
```

Comment:
This is the "one simple trick" that allows us to compare seamlessly between the string-encoded types that were parsed and the higher-order inferred ones (since we can reduce both to the same string representation). 

However, I'm not really comfortable with this because it is too limiting. In my opinion we do need to parse the defined types properly, but I'm not sure if this is in the scope of this PR. Thoughts?

Reply:
Same thoughts here. I think parsing the types into their respective data structures is definitely the scope of the PR. Like we talked about before, the string comparison is not going to work for union types where order does not matter, but I'm okay with tackling that part in a different PR because we can definitely live with a strict "ordered" type checking until we properly implement "is_subtype?(type, type_to_check)" for checking type compatibility. What do you think?

Reply:
I'll tackle this in this PR then. I'll first move the basic type conversions to the parser first, and then handle these cases where we need to convert from string to the type struct.